feat(CLI): allow for custom plugins

willemneal opened the pull request:
### What

Now any binary with the prefix `soroban-`, e.g. soroban-plugin will be executed with `soroban plugin`. Similar to `cargo-`.

Also `soroban --list` will list the installed plugins.

### Why

Allows devs to create plugins. E.g. `soroban network` could be used for launching local nodes.

### Known limitations

[TODO or N/A]

paulbellamy commented on cmd/soroban-cli/src/commands/custom.rs:
```diff
@@ -0,0 +1,29 @@
+    MissingSubcommand,
+    #[error(transparent)]
+    IO(#[from] std::io::Error),
+    #[error("Cannot find executable: {0}")]
```
Something like the cargo errors here might be nice:
```
$ cargo whatever
error: no such subcommand: `whatever`

     View all installed subcommands with `cargo --list`
```

Then `cargo --list` gives you a nice list of all the known subcommands.

willemneal replied:
Done! Even added a suggestion for closely matched words.

paulbellamy reviewed (commented):
Looks great so far, a couple questions.

willemneal commented on cmd/soroban-cli/src/bin/main.rs:
```diff
@@ -1,11 +1,18 @@
-        let mut cmd = Root::command();
-        e.format(&mut cmd).exit();
+    let root = Root::try_parse().unwrap_or_else(|e| {
+        if let clap::error::ErrorKind::UnknownArgument | clap::error::ErrorKind::InvalidSubcommand =
```
Yep!

willemneal replied:
fixed

willemneal commented on cmd/soroban-cli/src/commands/custom.rs:
Done.

willemneal commented on cmd/crates/soroban-test/tests/it/plugin.rs:
```diff
@@ -0,0 +1,35 @@
+        .assert()
+        .stdout("Hello, world!\n");
+
+    // Call soroban with the PATH variable set to include the target/bin directory
```
```suggestion
    // Call `soroban --list` with the PATH variable set to include the target/bin directory
```

paulbellamy commented on cmd/soroban-cli/src/bin/main.rs:
```diff
@@ -1,11 +1,23 @@
-        let mut cmd = Root::command();
-        e.format(&mut cmd).exit();
+    let root = Root::try_parse().unwrap_or_else(|e| {
+        if std::env::args().any(|s| &s == "--list") {
```
Suggestion: This seems like it ends up handling `--list` anywhere in the command:
e.g. `soroban contract --list`

Which is a bit surprising. Shouldn't this only be hit if there is no subcommand?

paulbellamy replied:
Like:
```diff
diff --git a/cmd/soroban-cli/src/bin/main.rs b/cmd/soroban-cli/src/bin/main.rs
index ea07f92..be9d2e1 100644
--- a/cmd/soroban-cli/src/bin/main.rs
+++ b/cmd/soroban-cli/src/bin/main.rs
@@ -4,10 +4,6 @@ use soroban_cli::{commands::plugin, Root};
 #[tokio::main]
 async fn main() {
     let root = Root::try_parse().unwrap_or_else(|e| {
-        if std::env::args().any(|s| &s == "--list") {
-            println!("{}", plugin::list().unwrap_or_default().join("\n"));
-            std::process::exit(0);
-        }
         if let clap::error::ErrorKind::InvalidSubcommand = e.kind() {
             if let Err(error) = plugin::run() {
                 eprintln!("error: {error}");
diff --git a/cmd/soroban-cli/src/commands/mod.rs b/cmd/soroban-cli/src/commands/mod.rs
index 829e82b..ee42c35 100644
--- a/cmd/soroban-cli/src/commands/mod.rs
+++ b/cmd/soroban-cli/src/commands/mod.rs
@@ -54,7 +54,7 @@ pub struct Root {
     #[arg(long)]
     pub list: bool,
     #[command(subcommand)]
-    pub cmd: Cmd,
+    pub cmd: Option<Cmd>,
 }
 
 impl Root {
@@ -72,12 +72,20 @@ impl Root {
     }
     pub async fn run(&self) -> Result<(), Error> {
         match &self.cmd {
-            Cmd::Contract(contract) => contract.run().await?,
-            Cmd::Events(events) => events.run().await?,
-            Cmd::Lab(lab) => lab.run().await?,
-            Cmd::Version(version) => version.run(),
-            Cmd::Completion(completion) => completion.run(),
-            Cmd::Config(config) => config.run()?,
+            Some(Cmd::Contract(contract)) => contract.run().await?,
+            Some(Cmd::Events(events)) => events.run().await?,
+            Some(Cmd::Lab(lab)) => lab.run().await?,
+            Some(Cmd::Version(version)) => version.run(),
+            Some(Cmd::Completion(completion)) => completion.run(),
+            Some(Cmd::Config(config)) => config.run()?,
+            None => {
+                if self.list {
+                    println!("{}", plugin::list().unwrap_or_default().join("\n"));
+                } else {
+                    let _ = Self::command().print_help();
+                    std::process::exit(1);
+                }
+            }
         };
         Ok(())
     }
```

willemneal replied:
Fixed it by checking if there is a missing subcommand.  Pushed, but still need to figure out the windows hex string issue.

paulbellamy commented on cmd/soroban-cli/src/commands/plugin.rs:
```diff
@@ -0,0 +1,79 @@
+        })
+        .filter(|s| !(is_hex_string(s) && s.len() > MAX_HEX_LENGTH))
+        .map(|s| s.replace("soroban-", ""))
+        .collect())
```
Seems like there's a bug in the hex processing on windows, because the test is failing

willemneal replied:
Yeah I need to check on this with a local windows test.

paulbellamy reviewed (commented):
Almost there, I think. One suggestion, one failing test on windows...

willemneal commented:
@paulbellamy I fixed the issue you raised with `--list` working anywhere.  Now it's just if there is a missing subcommand.

As for the windows issue, I changed the test to just check for the `hello` plugin.  I figure it's better to get the plugins in now and figure out the issue of an extra listed plugin in a future PR.
